src/uct/ib/ud/base/ud_ep.c

```diff
@@ -881,92 +1001,76 @@ static void uct_ud_ep_rx_ctl(uct_ud_iface_t *iface, uct_ud_ep_t *ep,
     uct_ud_ep_set_state(ep, UCT_UD_EP_FLAG_CREP_RCVD);
 }
 
-uct_ud_send_skb_t *uct_ud_ep_prepare_creq(uct_ud_ep_t *ep)
+static void uct_ud_ep_rx_ctl(uct_ud_iface_t *iface, uct_ud_ep_t *ep,
+                             uct_ud_neth_t *neth)
 {
-    uct_ud_iface_t *iface = ucs_derived_of(ep->super.super.iface, uct_ud_iface_t);
-    uct_ud_ctl_hdr_t *creq;
-    uct_ud_send_skb_t *skb;
-    uct_ud_neth_t *neth;
-    ucs_status_t status;
-
-    ucs_assert_always(ep->dest_ep_id == UCT_UD_EP_NULL_ID);
-    ucs_assert_always(ep->ep_id != UCT_UD_EP_NULL_ID);
-
-    /* CREQ should not be sent if CREP for the counter CREQ is scheduled
-     * (or sent already) */
-    ucs_assertv_always(!uct_ud_ep_ctl_op_check(ep, UCT_UD_EP_OP_CREP) &&
-                       !(ep->flags & UCT_UD_EP_FLAG_CREP_SENT),
-                       "iface=%p ep=%p conn_sn=%d rx_psn=%u ep_flags=0x%x "
-                       "ctl_ops=0x%x rx_creq_count=%d",
-                       iface, ep, ep->conn_sn, ep->rx.ooo_pkts.head_sn,
-                       ep->flags, ep->tx.pending.ops, ep->rx_creq_count);
-
-    skb = uct_ud_iface_get_tx_skb(iface, ep);
-    if (!skb) {
-        return NULL;
-    }
-
-    neth = skb->neth;
-    uct_ud_neth_init_data(ep, neth);
-
-    neth->packet_type  = UCT_UD_EP_NULL_ID;
-    neth->packet_type |= UCT_UD_PACKET_FLAG_CTL;
-
-    creq = (uct_ud_ctl_hdr_t *)(neth + 1);
-
-    creq->type                = UCT_UD_PACKET_CREQ;
-    creq->conn_req.conn_sn    = ep->conn_sn;
-    creq->conn_req.path_index = ep->path_index;
+    uct_ud_ctl_hdr_t *ctl = (uct_ud_ctl_hdr_t*)(neth + 1);
 
-    status = uct_ud_ep_get_address(&ep->super.super,
-                                   (void*)&creq->conn_req.ep_addr);
-    if (status != UCS_OK) {
-        return NULL;
-    }
+    ucs_trace_func("");
 
-    status = uct_ib_iface_get_device_address(&iface->super.super.super,
-                                             (uct_device_addr_t*)uct_ud_creq_ib_addr(creq));
-    if (status != UCS_OK) {
-        return NULL;
+    switch (ctl->type) {
+    case UCT_UD_PACKET_CREQ:
+        break;
+    case UCT_UD_PACKET_CREP:
+        uct_ud_ep_rx_crep(iface, ep, neth, ctl);
+        break;
+    case UCT_UD_PACKET_FIN:
+        /* Got FIN packet which means that the endpoint won't be used to
+         * to data anymore */
+        ucs_assert(!(ep->flags & UCT_UD_EP_FLAG_CONNECT_TO_EP));
+        uct_ud_ep_handle_peer_disconnect(ep);
+        break;
+    default:
+        ucs_fatal("unexpected control packet type - %d", ctl->type);
     }
+}
 
-    uct_ud_peer_name(ucs_unaligned_ptr(&creq->peer));
+static UCS_F_NOINLINE void
+uct_ud_ep_process_unexp_packet(uct_ud_iface_t *iface, uct_ud_ep_t *ep,
+                               uct_ud_neth_t *neth, unsigned length)
+{
+    char buf[256] = {0};
```

Comment:
why need to initialize? i would expect uct_ud_dump_packet to make sure it ends by \0

Reply:
we initiailize it in other places too when using `uct_ud_dump_packet`.
it doesn't add trailing `\0`.